### PR TITLE
fix(context): ensures all fields in Context are picklable

### PR DESCRIPTION
**Pull request by mabdinur**
Resolves: https://github.com/DataDog/dd-trace-py/issues/7849

We will need to make baggage picklable as well. 

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: tests/tracer/test_context.py [hunk 1]
```diff
@@ -88,6 +89,24 @@ def validate_traceparent(context, sampled_expected):
...
+                SpanLink(
+                    trace_id=123 << 64,
+                    span_id=456,
+                    tracestate="congo=t61rcWkgMzE",
+                    flags=1,
+                    attributes={"link.name": "some_name"},
+                )
+            ],
+        ),
+        Context(
```
Comment (emmettbutler) [inline tests/tracer/test_context.py:107]: Does this test fail without the functional change? When I tried adding a similar test case I couldn't get it to fail.

**Review by emmettbutler — APPROVED** [review]

**Review by majorgreys — CHANGES_REQUESTED** [review]
Add test run before and after fix.

See https://github.com/DataDog/dd-trace-py/pull/7949#discussion_r1433032067

File: tests/tracer/test_context.py (hunk 1 above)
Comment (mabdinur) [reply tests/tracer/test_context.py:107]: No I think the test will always pass. I'll modify. 

You need to access the unserialized fields on the context object to raise the attribute error.

Comment (mabdinur) [reply tests/tracer/test_context.py:107]: @emmettbutler I found the issue: 6c7aa449227ee66d0c64564c2dc0df03b8cfdce9. The test was not checking the span_links attribute

**Comment by mabdinur** [comment]
> Add test run before and after fix.
> 
> See [#7949 (comment)](https://github.com/DataDog/dd-trace-py/pull/7949#discussion_r1433032067)

Failing test: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/52865/workflows/a5f1da56-cb50-413d-bd20-37f11cf9546a/jobs/3366077

Passing test: https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/52866/workflows/16b4aa7c-0059-43fd-a936-a25bc3cbfedb

File: ddtrace/context.py [hunk 2]
```diff
@@ -246,17 +250,20 @@ def __eq__(self, other):
...
                     and self._metrics == other._metrics
+                    and self._span_links == other._span_links
+                    and self._baggage == other._baggage
                 )
         return False
 
     def __repr__(self):
         # type: () -> str
-        return "Context(trace_id=%s, span_id=%s, _meta=%s, _metrics=%s, _span_links=%s)" % (
+        return "Context(trace_id=%s, span_id=%s, _meta=%s, _metrics=%s, _span_links=%s _baggage=%s)" % (
```
Comment (mabdinur) [inline ddtrace/context.py:260]:
```suggestion
        return "Context(trace_id=%s, span_id=%s, _meta=%s, _metrics=%s, _span_links=%s, _baggage=%s)" % (
```

**Review by majorgreys — APPROVED** [review]

**Pull request merged**